sql/distsqlrun: refactor tableReader to implement RowSource

petermattis opened the pull request:
Refactor tableReader to implement the RowSource interface. Refactor
tableReader.Run() to be implemented in terms of
tableReader.Next() (i.e. the RowSource interface).
See #20550

Release note: None

petermattis committed:
sql/distsqlrun: refactor tableReader to implement RowSource

Refactor tableReader to implement the RowSource interface. Refactor
tableReader.Run() to be implemented in terms of
tableReader.Next() (i.e. the RowSource interface).

Adjusted BenchmarkTableReader to avoid using a RowBuffer. This shows the
benefit that can be achieved by using TableReader as a RowSource ("old"
below is with the benchmark modified to use a RowChannel).

name           old time/op  new time/op  delta
TableReader-8  11.6ms ± 5%   9.4ms ± 3%  -18.81%  (p=0.000 n=10+10)

See #20550

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20584)
<!-- Reviewable:end -->

RaduBerinde commented:
I will reiterate that I believe we should first work on row batching before going down this path because 1) it is definitely necessary regardless of other improvements and 2) it could reduce the benefit of this kind of change. Overall the change itself seems fine but if we make more changes to remove a row channel here and there, things could easily become very messy.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L-sD6A5GexaewqwiAi5:bvx6qa0)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
+1 to batching first :)

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L-sOAg0sMjNNc1mfPW6:b-n6wl6h)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
And there's another thing - there's been a series of triparty talks between
Radu, Rafa and myself about supporting synchronous scheduling of co-located
processors. You'd probably end up with something similar to what you have
here, but hopefully done in a more general way. I was hoping that line of
thought would marinate as we work towers unifying the execution engines.

BTW, I applaud the development of benchmarks like the one that lead to this
PR.

On Dec 8, 2017 6:11 PM, "Andrei Matei" <notifications@github.com> wrote:

> +1 to batching first :)
> ------------------------------
>
> Review status: 0 of 2 files reviewed at latest revision, all discussions
> resolved, some commit checks failed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L-sOAg0sMjNNc1mfPW6:b-n6wl6h>*
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/20584#issuecomment-350395250>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcWhUw0482VhZl7-NtxD7npCfM5Ujks5s-cIIgaJpZM4Q7tAB>
> .
>

petermattis commented:
Note that I hacked this up to see what could be accomplished. I'm not in anyway proposing this as the right approach. Passing batches of rows will lessen the need for removing inter-processor communication speed, but not eliminate it. And this PR gives us an indication of the possible speed-up from row batching (assuming we don't change processor internals at the same time). 

@andreimatei How would synchronous scheduling of co-located processors work? Would it be any different from pushing this PR to its conclusion and eliding `RowChannel` between co-located processors whenever possible? Probably better to discuss on #20550.

arjunravinarayan commented:
Another thing to note is that making the underlying MVCCScan go much faster probably exacerbates the difference in performance between the two approaches, making the need for eliding `RowChannel`s where possible more critical.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L02f41718W867x8S0Fg:b-1jq8ny)*
<!-- Sent from Reviewable.io -->

petermattis commented:
This is prep work for #20550, without actually taking the step of eliding RowChannels. The plan is to follow-up with similar refactorings of the other processors. Note the first commit is all context plumbing work. PTAL.

---

Review status: 0 of 26 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/distsqlrun/tablereader.go, line 374 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L0Gm62z4WtVP7U6GHrU:-L0Gm62z4WtVP7U6GHrV:b6qm8us) ([raw file](https://github.com/cockroachdb/cockroach/blob/f309fc90cb08f0d762ade449bc430e3d3bf84e4f/pkg/sql/distsqlrun/tablereader.go#L374)):*
> ```Go
> 
> func (tr *tableReader) ConsumerDone() {
> 	// TODO(peter): what to do here?
> ```

@raduberinde Any insight on what to do here? The various `RowSource` implementations which do something for this method only utilize the effects of that something in `RowReceiver.Push`. So it seems fine to do nothing here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L0GmXn3LlSra773W_d-:b-adyzgd)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 26 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/tablereader.go, line 314 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L0H7gIBQce1k6azIbfg:-L0H7gIBQce1k6azIbfh:bhc7qdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f309fc90cb08f0d762ade449bc430e3d3bf84e4f/pkg/sql/distsqlrun/tablereader.go#L314)):*
> ```Go
> }
> 
> func (tr *tableReader) Next(ctx context.Context) (sqlbase.EncDatumRow, ProducerMetadata) {
> ```

Currently the table reader calls `tr.sendMisplannedRangesMetadata` and `sendTraceData` before finishing. Without something similar here (returning that metadata after the last row) we wouldn't get traces or (more importantly) misplanned ranges info from the table reader.

---

*[pkg/sql/distsqlrun/tablereader.go, line 374 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L0Gm62z4WtVP7U6GHrU:-L0H6Y1GPoLYL3bgq3nO:b-6a05kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f309fc90cb08f0d762ade449bc430e3d3bf84e4f/pkg/sql/distsqlrun/tablereader.go#L374)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

@raduberinde Any insight on what to do here? The various `RowSource` implementations which do something for this method only utilize the effects of that something in `RowReceiver.Push`. So it seems fine to do nothing here. 
</blockquote></details>

Currently, even if a consumer signals that it is done, it still "drains" the input (e.g. to get any remaining metadata). If that carries over, the consumer will keep running Next, so the right thing to do would be to set a flag and have the next `Next` call say it's done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 26 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsqlrun/tablereader.go, line 314 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L0H7gIBQce1k6azIbfg:-L0HY7FzelT0plfgn1qX:bfi2bic) ([raw file](https://github.com/cockroachdb/cockroach/blob/f309fc90cb08f0d762ade449bc430e3d3bf84e4f/pkg/sql/distsqlrun/tablereader.go#L314)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Currently the table reader calls `tr.sendMisplannedRangesMetadata` and `sendTraceData` before finishing. Without something similar here (returning that metadata after the last row) we wouldn't get traces or (more importantly) misplanned ranges info from the table reader.
</blockquote></details>

Ah, I missed that. Is the contract that we'll send the misplanned ranges metadata and tracedata even if there is an error? Is there a reason this extra metadata isn't bundled into the same `ProducerMetadata` struct. Currently it looks like we send 2 separate ProducerMetadata.

PTAL at the logic both in `Next()` and `Run()`. There is lots of logic here which feels like it could be shared across processors. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 26 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsqlrun/tablereader.go, line 314 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L0H7gIBQce1k6azIbfg:-L0HpRMD_6XAd4HSEdFE:b7ra46b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f309fc90cb08f0d762ade449bc430e3d3bf84e4f/pkg/sql/distsqlrun/tablereader.go#L314)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ah, I missed that. Is the contract that we'll send the misplanned ranges metadata and tracedata even if there is an error? Is there a reason this extra metadata isn't bundled into the same `ProducerMetadata` struct. Currently it looks like we send 2 separate ProducerMetadata.

PTAL at the logic both in `Next()` and `Run()`. There is lots of logic here which feels like it could be shared across processors. 
</blockquote></details>

I think they can probably be packaged in the same `ProducerMetadata`.

I think we want the tracing data even if there's an error. Note that finishing early (through `ConsumerDone`) is not necessarily an error, it can happen because of a LIMIT (and we definitely want the metadata in that case).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 26 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/tablereader.go, line 314 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L0H7gIBQce1k6azIbfg:-L0HrGntawrQ-ZPvPZ3U:bioo7v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f309fc90cb08f0d762ade449bc430e3d3bf84e4f/pkg/sql/distsqlrun/tablereader.go#L314)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think they can probably be packaged in the same `ProducerMetadata`.

I think we want the tracing data even if there's an error. Note that finishing early (through `ConsumerDone`) is not necessarily an error, it can happen because of a LIMIT (and we definitely want the metadata in that case).
</blockquote></details>

Yeah, packaging everything into a single `ProducerMetadata` seems to work. See `tableReader.finalProducerMeta`.

The changes here have gotten complex enough that I'm not at all comfortable with them without additional testing. I'd like to add some testing which verifies both the `Processor` and `RowSource` interface behaviors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Just a meta-comment: I can easily see this series of changes turning into a can of worms and potentially causing a lot of conflicts with other ongoing work (row batching). Consider putting things on a separate branch for now (you can still post PRs for changes that go on that branch).

petermattis commented:
> Just a meta-comment: I can easily see this series of changes turning into a can of worms and potentially causing a lot of conflicts with other ongoing work (row batching). Consider putting things on a separate branch for now (you can still post PRs for changes that go on that branch).

That's a good point. I was hoping to do this processor-by-processor to avoid some uber merge nightmare. I think the row batching could take a similar approach.

petermattis commented:
Ok, I think I have all of the semantics of `RowSource` implemented correctly now. The restructuring is now much less pleasant to look at, but I think the final code is reasonable. The PR now takes @andreimatei's suggestion and embeds a context in `tableReader` for use by `Next()`. When used as a `RowSource`, `tableReader` creates a span on the first call to `Next()` and finishes the span either when it is done sending rows or when `ConsumerClosed()` is called.

At this point, it looks feasible to implement `tableReader.Run` as:

```
func runProcessor(p Processor, out RowReceiver, wg *sync.WaitGroup)
```

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsqlrun/flow.go, line 57 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L0Lj1f0b2-Elac-Mywb:-L0Lj1f0b2-Elac-Mywc:b-aaicqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d414383a361af1eff73e5749962fe580c5acdf7/pkg/sql/distsqlrun/flow.go#L57)):*
> ```Go
> 	// Context used for all execution within the flow.
> 	// Created in Start(), canceled in Cleanup().
> 	ctx context.Context
> ```

This was moved from `Flow.ctx` so that all processors have access to the flow's context. A follow-on change would be to remove the `ctx` parameter from `Processor.Run`.

---

*[pkg/sql/distsqlrun/tablereader.go, line 374 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L0Gm62z4WtVP7U6GHrU:-L0LiopOQRr2TjbGrBLq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f309fc90cb08f0d762ade449bc430e3d3bf84e4f/pkg/sql/distsqlrun/tablereader.go#L374)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Currently, even if a consumer signals that it is done, it still "drains" the input (e.g. to get any remaining metadata). If that carries over, the consumer will keep running Next, so the right thing to do would be to set a flag and have the next `Next` call say it's done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L0LirKeG8BVsjgmbUFF:bhev9fs)*
<!-- Sent from Reviewable.io -->

petermattis committed:
sql/distsqlrun: add generalized Run(RowSource, RowReceiver)

Implement `tableReader.Run()` using the generalized `Run()` function.

Release note: None

petermattis commented:
> At this point, it looks feasible to implement tableReader.Run as:

I'm not sure what that function signature was about, but the real signature is:

```
func Run(ctx context.Context, src RowSource, dst RowReceiver)
```

Seems to work and `tableReader.Run()` is now trivial. See the 2nd commit for details.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L0Lspv5R0NA09xPw1n0:b-75ypda)*
<!-- Sent from Reviewable.io -->

petermattis committed:
sql/distsqlrun: refactor valuesProcessor to implement RowSource

Release note: None

petermattis committed:
sql/distsqlrun: refactor distinct to implement RowSource

Release note: None

petermattis commented:
I've add two additional commits which refactor `valuesProcessor` and `distinct` to implement `RowSource`. The work now seems somewhat mechanical. There is definitely some common code that could be shared. For example, `processorBase` could implement `Types`, `ConsumerDone` and `ConsumerClosed`.

petermattis commented:
Ping. I'd like to get this reviewed and merged piecemeal rather than changing all processors at once. This PR changes 3 of the 15 existing processors to implement `RowSource`, but does not otherwise change the interaction between processors (i.e. it does not elide `RowChannel`). Changing the remaining processors should not be difficult.

---

Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L1Sh-GrT-PEAUXCGaNb:b-rwhvqk)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM

@andreimatei should take a look too. And folks who are working on row batching should at least take a look at the new interface. CC @arjunravinarayan @asubiotto

---

Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/tablereader.go, line 326 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1Ss_B0JcQB8esduBXH:-L1Ss_B0JcQB8esduBXI:b-aggseb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c0b6bd7277ce510f40ed6f4c15e0631d9ce7e/pkg/sql/distsqlrun/tablereader.go#L326)):*
> ```Go
> 		}
> 
> 		if tr.ctx == nil {
> ```

How can `tr.ctx` be anything but `nil` here? (given that `!tr.started`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L1StfqUvPooeiCJcX8A:bk4zukt)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsqlrun/tablereader.go, line 326 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1Ss_B0JcQB8esduBXH:-L1TAKFHnEMHDz0NIFKB:b-s7unw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c0b6bd7277ce510f40ed6f4c15e0631d9ce7e/pkg/sql/distsqlrun/tablereader.go#L326)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

How can `tr.ctx` be anything but `nil` here? (given that `!tr.started`)
</blockquote></details>

It can't. I think this was detritus from earlier versions of this change (before I added `tr.started`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/tablereader.go, line 57 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TS_jkWPiSAgdCjkHa:-L1TS_jlmMY2XDHzCUJ8:b692lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L57)):*
> ```Go
> 	processorBase
> 
> 	ctx     context.Context
> ```

please comment how `ctx` should be used instead of `flowCtx.Ctx`. 

---

*[pkg/sql/distsqlrun/tablereader.go, line 201 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TaudT4ArVnVAdClOz:-L1TW-bjy7BeZTKkQ1MZ:b62v2iz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L201)):*
> ```Go
> func (tr *tableReader) Run(ctx context.Context, wg *sync.WaitGroup) {
> 	if tr.out.output == nil {
> 		panic("output RowReceiver not initialized for emitting rows")
> ```

this message is OK in `ProcOutputHelper`, but it's not great here. Just say "output not configured". Or, better yet, move the assertion to `newTableReader()`.

---

*[pkg/sql/distsqlrun/tablereader.go, line 204 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TVPNxAx3-vN3dClT8:-L1TVPNxAx3-vN3dClT9:b2br8nh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L204)):*
> ```Go
> 	}
> 	if ctx != tr.flowCtx.ctx {
> 		panic("unexpected context")
> ```

please add a TODO somewhere around here to remove the `ctx` arg. Or ideally do it now. Otherwise I don't really know what's going on here.

---

*[pkg/sql/distsqlrun/tablereader.go, line 229 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TakwJGqhkp1eSQTrp:-L1TakwJGqhkp1eSQTrq:ba0ciyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L229)):*
> ```Go
> 
> 	if tr.consumerStatus != ConsumerClosed {
> 		if meta := tr.producerMeta(nil); !meta.Empty() {
> ```

pls comment nils (some of them were commented before :) )

---

*[pkg/sql/distsqlrun/tablereader.go, line 312 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TbbYO8Wr8JYxaRoCP:-L1TbbYO8Wr8JYxaRoCQ:bogolmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L312)):*
> ```Go
> 	if tr.ctx != nil {
> 		if log.V(1) {
> 			log.Infof(tr.ctx, "exiting")
> ```

`VEventf()` instead of `if log.v(1)` please. I know it was bad before too.

---

*[pkg/sql/distsqlrun/tablereader.go, line 315 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TboyFtrhRHp33OKd6:-L1TboyFtrhRHp33OKd7:bl6qry2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L315)):*
> ```Go
> 		}
> 		tracing.FinishSpan(tr.span)
> 		tr.ctx, tr.span = nil, nil
> ```

not sure resetting these is a good idea (it forces you to not use `tr.ctx` in other places although I think you could even with the finished span).

---

*[pkg/sql/distsqlrun/tablereader.go, line 324 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1Tah4ho6pvUNGjAAH3:-L1Tah4ho6pvUNGjAAH4:b-v4e3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L324)):*
> ```Go
> // terminated, either due to being indicated by the consumer, or because the
> // tableReader ran out of rows or encountered an error.
> func (tr *tableReader) producerMeta(err error) ProducerMetadata {
> ```

pls comment that err can be nil

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1TcDm2wXRG6MAAJOjn:b-31of62) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
> ```Go
> 	var meta ProducerMetadata
> 	if tr.ctx != nil {
> 		meta = ProducerMetadata{
> ```

`ProducerMetadata` is specified to have a single fields set and has the following message:
	// Only one of these fields will be set. If this ever changes, note that
	// there's consumers out there that extract the error and, if there is one,
	// forward it in isolation and drop the rest of the record.

---

*[pkg/sql/distsqlrun/tablereader.go, line 339 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TZ8OOODVw0CnAMt7L:-L1TZ8OOODVw0CnAMt7M:b4mphg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L339)):*
> ```Go
> }
> 
> func (tr *tableReader) Types() []sqlbase.ColumnType {
> ```

how come the linter is not  yelling at you?
Please comment that these methods are "part of the RowSource interface".

---

*[pkg/sql/distsqlrun/tablereader.go, line 343 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TZwLG7aH_Die0V3W_:-L1TZwLG7aH_Die0V3Wa:b-w1joh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L343)):*
> ```Go
> }
> 
> func (tr *tableReader) Next() (sqlbase.EncDatumRow, ProducerMetadata) {
> ```

With processors implementing the `RowSource` iface, the comment on the `Next` method is no longer appropriate - it talks about rows pushed by a producer; it should be expanded.
And it should also say something about post-processing - that the processors have to do their post-processing before returning a row.

---

*[pkg/sql/distsqlrun/tablereader.go, line 418 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1T_xJbL88TsO8jPCm9:-L1T_xJbL88TsO8jPCmA:b-bb0ngd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L418)):*
> ```Go
> func (tr *tableReader) ConsumerDone() {
> 	if tr.consumerStatus != NeedMoreRows {
> 		log.Fatalf(context.Background(), "tableReader already done or closed: %d", tr.consumerStatus)
> ```

does it not work if you use tr.ctx (even if its span has been closed already)? I seem to remember that we made it logging to finished spans "work" (not panic).

---

*[pkg/sql/distsqlrun/tablereader_test.go, line 233 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TUJYo64ra3HoTT4b5:-L1TUJYo64ra3HoTT4b6:b2pnift) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader_test.go#L233)):*
> ```Go
> 				out = tr
> 			} else {
> 				tr.Run(context.Background(), nil)
> ```

nit: you've done a superfluous change to this line, but you missed the opportunity to comment the nil :P

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/tablereader.go, line 57 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TS_jkWPiSAgdCjkHa:-L1Te30TtJKfAML9Y5AW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L57)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please comment how `ctx` should be used instead of `flowCtx.Ctx`. 
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/tablereader.go, line 201 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TaudT4ArVnVAdClOz:-L1TegkexYvjtLyaPwPj:btdfd4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L201)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this message is OK in `ProcOutputHelper`, but it's not great here. Just say "output not configured". Or, better yet, move the assertion to `newTableReader()`.
</blockquote></details>

Fixed the message. `tablereader_test.go` creates table readers with a `nil` output. Agreed this should be cleaned up, but we're not yet at that point.

---

*[pkg/sql/distsqlrun/tablereader.go, line 204 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TVPNxAx3-vN3dClT8:-L1TeM9Rn1r6nARqRp10:b-v6akik) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L204)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a TODO somewhere around here to remove the `ctx` arg. Or ideally do it now. Otherwise I don't really know what's going on here.
</blockquote></details>

I'll remove it in a follow-on PR. I pinky promise.

---

*[pkg/sql/distsqlrun/tablereader.go, line 229 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TakwJGqhkp1eSQTrp:-L1Tf-QF_YijTNVDcBTH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L229)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls comment nils (some of them were commented before :) )
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/tablereader.go, line 312 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TbbYO8Wr8JYxaRoCP:-L1Tf6yU8CC5tHF1P6l1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L312)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`VEventf()` instead of `if log.v(1)` please. I know it was bad before too.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/tablereader.go, line 315 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TboyFtrhRHp33OKd6:-L1Tf8D-pyMwRZQ_V17F:b2t8vlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L315)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

not sure resetting these is a good idea (it forces you to not use `tr.ctx` in other places although I think you could even with the finished span).
</blockquote></details>

Well, we use `tr.started && tr.ctx == nil` to indicate closed. I could certainly do something else here, though I'm not sure if it would be an improvement. Do you have a more specific reason why you're concerned about setting `tr.{ctx,span}` to `nil`?

---

*[pkg/sql/distsqlrun/tablereader.go, line 324 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1Tah4ho6pvUNGjAAH3:-L1Tf_1MurLvItJdYLdy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L324)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls comment that err can be nil
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1TgCqmxCPvHCjGr3uT:b-w03ehz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`ProducerMetadata` is specified to have a single fields set and has the following message:
	// Only one of these fields will be set. If this ever changes, note that
	// there's consumers out there that extract the error and, if there is one,
	// forward it in isolation and drop the rest of the record.
</blockquote></details>

Hmm, let me investigate this again. I think I took a look while developing this PR and determined that the comment was out of date.

---

*[pkg/sql/distsqlrun/tablereader.go, line 339 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TZ8OOODVw0CnAMt7L:-L1TftQ-myaU6aJ4ER0k:bk2nvt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L339)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how come the linter is not  yelling at you?
Please comment that these methods are "part of the RowSource interface".
</blockquote></details>

Done. The linter only cares about exported types.

---

*[pkg/sql/distsqlrun/tablereader.go, line 343 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TZwLG7aH_Die0V3W_:-L1TgR224RHXfXHdFdbS:bdfwngx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L343)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

With processors implementing the `RowSource` iface, the comment on the `Next` method is no longer appropriate - it talks about rows pushed by a producer; it should be expanded.
And it should also say something about post-processing - that the processors have to do their post-processing before returning a row.
</blockquote></details>

I think I already updated the comment on `RowSource.Next`. Did I miss something that should be adjusted?

---

*[pkg/sql/distsqlrun/tablereader.go, line 418 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1T_xJbL88TsO8jPCm9:-L1Tg_OyvoVzqx_5PtAi:b-rrbq6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L418)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

does it not work if you use tr.ctx (even if its span has been closed already)? I seem to remember that we made it logging to finished spans "work" (not panic).
</blockquote></details>

Well, if we set `tr.ctx` to `nil` this won't work. I don't see a problem with using `context.Background()` here given that this is a fatal error.

---

*[pkg/sql/distsqlrun/tablereader_test.go, line 233 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TUJYo64ra3HoTT4b5:-L1Th0H48oSZgDGHBHQy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader_test.go#L233)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: you've done a superfluous change to this line, but you missed the opportunity to comment the nil :P
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1Tms7cmmUtaoFXWvoS:bkqu4xa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Hmm, let me investigate this again. I think I took a look while developing this PR and determined that the comment was out of date.
</blockquote></details>

I was wrong. There are several places in the code which extract the error and drop the rest of the metadata. `NoMetadataRowSource.NextRow` is the primary one (used by several processors) and other processors do this manually. Fixing `NoMetadataRowSource` is rather straightforward, but would it be better to leave the invariant in place and produce multiple bits of metadata here? That appears awkward, though perhaps I just need to stare at the code longer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1ToPriHj_MuQarlnPv:b8sx6w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I was wrong. There are several places in the code which extract the error and drop the rest of the metadata. `NoMetadataRowSource.NextRow` is the primary one (used by several processors) and other processors do this manually. Fixing `NoMetadataRowSource` is rather straightforward, but would it be better to leave the invariant in place and produce multiple bits of metadata here? That appears awkward, though perhaps I just need to stare at the code longer.
</blockquote></details>

Note that the over-the-wire proto is `RemoteProducerMetadata` which uses `oneof`. So if we support multiple things per `ProducerMetadata`, we will likely need to adjust the stream_encoder code to split it up into multiple protos.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1TolhDQucMmRR6iM8G:b-6k21dz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Note that the over-the-wire proto is `RemoteProducerMetadata` which uses `oneof`. So if we support multiple things per `ProducerMetadata`, we will likely need to adjust the stream_encoder code to split it up into multiple protos.
</blockquote></details>

I am inclined to believe that "one at a time, and nothing comes after an error" is an easy to work with convention, but I don't have strong feelings.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1Tp9Sd6_d-F_USctLY:b-lsmbo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I am inclined to believe that "one at a time, and nothing comes after an error" is an easy to work with convention, but I don't have strong feelings.
</blockquote></details>

Ok, so if an error is returned we don't need to return anything else. And if no error is returned I have to return a separate `ProducerMetadata` for each field set? Would be convenient to send the trace and misplanned ranges data at the same time, but I think that can be handled.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1U-A_ph0RXwuFxp-a3:b-2n3w7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ok, so if an error is returned we don't need to return anything else. And if no error is returned I have to return a separate `ProducerMetadata` for each field set? Would be convenient to send the trace and misplanned ranges data at the same time, but I think that can be handled.
</blockquote></details>

Ok, I took a crack at this. It mostly works, though causes `TestTrace` to fail deterministically. I need to investigate, but won't be able to do so until later tonight or tomorrow. PTAL and let me know if this looks better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/base.go, line 139 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1U7M5MRc0hZsyUG0_-:-L1U7M5MRc0hZsyUG0_0:b-3iuwjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428e5eb55d58a2559f246211d03f1785762d4d9/pkg/sql/distsqlrun/base.go#L139)):*
> ```Go
> 	ConsumerDone()
> 
> 	// ConsumerClosed informs the source that the consumer it is done and will
> ```

s/the consumer it is done/the consumer, well, it is done

---

*[pkg/sql/distsqlrun/tablereader.go, line 315 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TboyFtrhRHp33OKd6:-L1UD2_fQOE59QDFsDHU:braus29) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L315)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Well, we use `tr.started && tr.ctx == nil` to indicate closed. I could certainly do something else here, though I'm not sure if it would be an improvement. Do you have a more specific reason why you're concerned about setting `tr.{ctx,span}` to `nil`?
</blockquote></details>

Well, if `tr.ctx == nil` means something, then please comment it.
I don't like it very much because it's weird. Dealing with nil ctxs is not something we do. And that use of `context.Background`, consequential or not, also bothered me. At the very least because I had to figure out why we're doing it - and I guess I misunderstood too; I thought you were concerned about a finished span.

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1UAJFdfEzca_H85qT6:bm5xot2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ok, I took a crack at this. It mostly works, though causes `TestTrace` to fail deterministically. I need to investigate, but won't be able to do so until later tonight or tomorrow. PTAL and let me know if this looks better.
</blockquote></details>

I think I might have spoken nonsense before. A producer can push other metadata after an error, and so can a RowSource return metadata after an error. Sorry. And it's quite useful, e.g. for sending the trace after an error in this processor.
So I think the `tr.close()` call here should go away.
Perhaps that's also why that test is failing.

---

*[pkg/sql/distsqlrun/tablereader.go, line 339 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TZ8OOODVw0CnAMt7L:-L1U6VOqhVY9TS5V5cp_:bcz8mo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L339)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Done. The linter only cares about exported types.
</blockquote></details>

ignore: I personally have moved to the "Types is part of the RowSource interface" phrasing for interfaces with more than one method. The current phrasing doesn't quite compute.

---

*[pkg/sql/distsqlrun/tablereader.go, line 343 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TZwLG7aH_Die0V3W_:-L1U4tTm3o5sKcoyUnUI:bq6po49) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L343)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think I already updated the comment on `RowSource.Next`. Did I miss something that should be adjusted?
</blockquote></details>

Ah, you did change something indeed, that's good.
The thing that made me say something here is that I was confused for a bit about why we're not using `EmitRow()` any more. `EmitRow()` did two things - post-processing and sending to an output. Post-processing is now done in `Next()`. So I was thinking that a comment somewhere should say that. At first I thought the comment should be on `tableReader.Next()`; then I thought maybe a nod to post-processing belongs in `RowSource.Next()`. I dunno. You can also ignore.

---

*[pkg/sql/distsqlrun/tablereader.go, line 73 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1U8fl0-5CDrrRD6jSp:-L1U8fl0-5CDrrRD6jSq:b-f7jpcy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6429d869d92dbff53593ee2e4503ef99268dcf90/pkg/sql/distsqlrun/tablereader.go#L73)):*
> ```Go
> 
> 	started    bool
> 	halfClosed bool
> ```

If this stays, please put some comment on this field.
Feel free to ignore, but I find the way you've structured this draining a bit confusing. I would have created a buffer of metadata (say, `trailingMeta`) to send, and then, if `consumerStatus != NeedsMoreRows`, I would have made `Next()` return from that buffer until it's exhausted. I'm pretty sure there's precedent for such a "pattern" in various other places in DistSQL too.

---

*[pkg/sql/distsqlrun/tablereader.go, line 292 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1U9aPWEMnqP0hg6Oz2:-L1U9aPWEMnqP0hg6Oz3:bq2tx6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6429d869d92dbff53593ee2e4503ef99268dcf90/pkg/sql/distsqlrun/tablereader.go#L292)):*
> ```Go
> 	tr.halfClosed = true
> 	// This prevents Next() from returning more rows.
> 	tr.out.rowIdx = tr.out.maxRowIdx
> ```

Ummm seems like someone is abusing something here, and also breaking the `ProcOutputHelper`'s encapsulation.
Can we not look at `tr.halfClosed` instead of `tr.out.rowIdx` in `Next()`?
Or better yet I hope `halfClosed` goes away and we look at whether the `trailingMeta` buffer that I was suggesting elsewhere is non-nil.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->

petermattis commented:
TFTR! PTAL.

---

Review status: 0 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/distsqlrun/base.go, line 139 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1U7M5MRc0hZsyUG0_-:-L1Xy0zPVJn1_xA2F3LD:b-h9pbwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428e5eb55d58a2559f246211d03f1785762d4d9/pkg/sql/distsqlrun/base.go#L139)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/the consumer it is done/the consumer, well, it is done
</blockquote></details>

Done, though I just removed the extraneous `it`.

---

*[pkg/sql/distsqlrun/tablereader.go, line 315 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TboyFtrhRHp33OKd6:-L1Y2x9L7_I3e0Rpx4La:b-grz9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L315)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Well, if `tr.ctx == nil` means something, then please comment it.
I don't like it very much because it's weird. Dealing with nil ctxs is not something we do. And that use of `context.Background`, consequential or not, also bothered me. At the very least because I had to figure out why we're doing it - and I guess I misunderstood too; I thought you were concerned about a finished span.
</blockquote></details>

I've added an explicit `tr.closed` field and left `tr.ctx` alone. I'm still using `context.Background()` for the log fatals because that code is essentially untested (we never hit that code path in practice) and I'd prefer not to have surprises about using an invalid context.

---

*[pkg/sql/distsqlrun/tablereader.go, line 327 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TcDm2wXRG6MAAJOjm:-L1Y3AS6SRWW9I_2EqQj:brir6xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L327)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think I might have spoken nonsense before. A producer can push other metadata after an error, and so can a RowSource return metadata after an error. Sorry. And it's quite useful, e.g. for sending the trace after an error in this processor.
So I think the `tr.close()` call here should go away.
Perhaps that's also why that test is failing.

</blockquote></details>

I believe the test was failing because there were no misplanned ranges and thus we were sending an empty `ProducerMetadata` before sending the trace data.

---

*[pkg/sql/distsqlrun/tablereader.go, line 339 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TZ8OOODVw0CnAMt7L:-L1XyUifG0zF_ZNL46_S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L339)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ignore: I personally have moved to the "Types is part of the RowSource interface" phrasing for interfaces with more than one method. The current phrasing doesn't quite compute.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/tablereader.go, line 343 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TZwLG7aH_Die0V3W_:-L1Xy_yurEpqFXUDPNw4:b-4hk260) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L343)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Ah, you did change something indeed, that's good.
The thing that made me say something here is that I was confused for a bit about why we're not using `EmitRow()` any more. `EmitRow()` did two things - post-processing and sending to an output. Post-processing is now done in `Next()`. So I was thinking that a comment somewhere should say that. At first I thought the comment should be on `tableReader.Next()`; then I thought maybe a nod to post-processing belongs in `RowSource.Next()`. I dunno. You can also ignore.
</blockquote></details>

I'm not sure if anything needs to be commented here. Clearly post-processing needs to be done. Failure to do so will break tests dramatically.

---

*[pkg/sql/distsqlrun/tablereader.go, line 73 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1U8fl0-5CDrrRD6jSp:-L1Y2p55o6adEJbcOMNe:bm65881) ([raw file](https://github.com/cockroachdb/cockroach/blob/6429d869d92dbff53593ee2e4503ef99268dcf90/pkg/sql/distsqlrun/tablereader.go#L73)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

If this stays, please put some comment on this field.
Feel free to ignore, but I find the way you've structured this draining a bit confusing. I would have created a buffer of metadata (say, `trailingMeta`) to send, and then, if `consumerStatus != NeedsMoreRows`, I would have made `Next()` return from that buffer until it's exhausted. I'm pretty sure there's precedent for such a "pattern" in various other places in DistSQL too.
</blockquote></details>

Done. `halfClosed` has been replaced with `trailingMetadata`. Note that I didn't do this for `distinct` and `valuesProcessor` because their handling of trailing metadata is simpler.

---

*[pkg/sql/distsqlrun/tablereader.go, line 292 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1U9aPWEMnqP0hg6Oz2:-L1Xz1UNQQZqGMqCiwxx:b-esgxwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6429d869d92dbff53593ee2e4503ef99268dcf90/pkg/sql/distsqlrun/tablereader.go#L292)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Ummm seems like someone is abusing something here, and also breaking the `ProcOutputHelper`'s encapsulation.
Can we not look at `tr.halfClosed` instead of `tr.out.rowIdx` in `Next()`?
Or better yet I hope `halfClosed` goes away and we look at whether the `trailingMeta` buffer that I was suggesting elsewhere is non-nil.

</blockquote></details>

Yes, there is a bit of encapsulation breakage here, though I'm hoping it is temporary. Once all processors implement `RowSource`, we can start trimming stuff from `ProcOutputHelper` and the encapsulation breakage can be fixed. I made this mildly cleaner by adding a method to `ProcOutputHelper`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L1Y5NksMzoxL8BQNgA3:bousnez)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/base.go, line 139 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1U7M5MRc0hZsyUG0_-:-L1YJKlwPtIVGyDU4EH2:btdnf1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428e5eb55d58a2559f246211d03f1785762d4d9/pkg/sql/distsqlrun/base.go#L139)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Done, though I just removed the extraneous `it`.
</blockquote></details>

I kid.

---

*[pkg/sql/distsqlrun/tablereader.go, line 315 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1TboyFtrhRHp33OKd6:-L1YLNYKsjtBnZX6BfxT:bw2lcdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca6faa8e19bf641990aecde2e8ff2f157f23981/pkg/sql/distsqlrun/tablereader.go#L315)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I've added an explicit `tr.closed` field and left `tr.ctx` alone. I'm still using `context.Background()` for the log fatals because that code is essentially untested (we never hit that code path in practice) and I'd prefer not to have surprises about using an invalid context.
</blockquote></details>

Thanks. SGTM about about Background(); I now think it might not work with a closed span if we have lightstep enabled.

---

*[pkg/sql/distsqlrun/tablereader.go, line 73 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-L1U8fl0-5CDrrRD6jSp:-L1YJh43WEyjCIIfo1sZ:b-7rty62) ([raw file](https://github.com/cockroachdb/cockroach/blob/6429d869d92dbff53593ee2e4503ef99268dcf90/pkg/sql/distsqlrun/tablereader.go#L73)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Done. `halfClosed` has been replaced with `trailingMetadata`. Note that I didn't do this for `distinct` and `valuesProcessor` because their handling of trailing metadata is simpler.
</blockquote></details>

danke

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584#-:-L1YMxRbbR3G3VsOYiDQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@arjunravinarayan, @asubiotto I'm going to merge this now. Feel free to leave additional comments which I'll address in follow-on PRs. I don't think this fundamentally affects row batching, but if it does I'm happy to deal with the fallout.

arjunravinarayan commented:
Reviewed 1 of 26 files at r4, 4 of 4 files at r9, 3 of 3 files at r10, 2 of 2 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20584)*
<!-- Sent from Reviewable.io -->
